crates/ty_server/src/session/client.rs

```diff
@@ -47,63 +45,49 @@ impl Client {
     ) where
         R: lsp_types::request::Request,
     {
-        let response_handler = Box::new(move |client: &Client, response: lsp_server::Response| {
-            let _span =
-                tracing::debug_span!("client_response", id=%response.id, method = R::METHOD)
-                    .entered();
-
-            match (response.error, response.result) {
-                (Some(err), _) => {
-                    tracing::error!(
-                        "Got an error from the client (code {code}, method {method}): {message}",
-                        code = err.code,
-                        message = err.message,
-                        method = R::METHOD
-                    );
-                }
-                (None, Some(response)) => match serde_json::from_value(response) {
-                    Ok(response) => response_handler(client, response),
-                    Err(error) => {
-                        tracing::error!(
-                            "Failed to deserialize client response (method={method}): {error}",
-                            method = R::METHOD
-                        );
-                    }
-                },
-                (None, None) => {
-                    if TypeId::of::<R::Result>() == TypeId::of::<()>() {
-                        // We can't call `response_handler(())` directly here, but
-                        // since we _know_ the type expected is `()`, we can use
-                        // `from_value(Value::Null)`. `R::Result` implements `DeserializeOwned`,
-                        // so this branch works in the general case but we'll only
-                        // hit it if the concrete type is `()`, so the `unwrap()` is safe here.
-                        response_handler(client, serde_json::from_value(Value::Null).unwrap());
-                    } else {
-                        tracing::error!(
-                            "Invalid client response: did not contain a result or error (method={method})",
-                            method = R::METHOD
-                        );
-                    }
-                }
-            }
-        });
+        self.send_request_raw(
+            session,
+            SendRequest {
+                method: R::METHOD.to_string(),
+                params: serde_json::to_value(params).expect("Params to be serializable"),
+                response_handler: ClientResponseHandler::for_request::<R>(response_handler),
+            },
+        );
+    }
+
+    pub(crate) fn send_deferred_request<R>(
```

Comment:
I'm still a bit unsure on why this is required. Do you mind explaining in brief on the reason this is required?

Reply:
`send_request` requires a `&Session`. We don't have a `&Session` in the background worker. We only have a `SessionSnapshot`.